File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/MakeTempCardInDiscardAction.ts
import Async from "../../../../../Core/Async";
import AbstractCard from "../../Cards/AbstractCard";
import EffectKit from "../../Effect/EffectKit";
import { ShowCardAndAddToDiscardEffect } from "../../Effect/ShowCardAndAddToDiscardEffect";
import AbstractGameAction, { ActionType } from "../AbstractGameAction";

export class MakeTempCardInDiscardAction extends AbstractGameAction
{
    public constructor(private card : AbstractCard, private amount = 1, private sameUUID = false) {
        super();
        this.Type = ActionType.CARD_MANIPULATION; 
    }

    async Update(): Promise<boolean> {
        for (let i = 0; i < this.amount; i++) {
            EffectKit.Inst().Play(new ShowCardAndAddToDiscardEffect(this.card.Clone()))
        }
        await Async.Delay(0.5);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Combat/MakeTempCardInDiscardAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Card;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Event;

namespace ModTheSpire.Effect.Other
{
    public class MakeTempCardInDiscardAction : IGameAction
    {
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.CARD_MANIPULATION;
        private int mNumCards;
        private AbstractCard mCard;
        private bool mSameUUID;
        public MakeTempCardInDiscardAction(AbstractCard card, int amount = 1,bool sameUUID = false) {
            mNumCards = amount;
            mCard = card;
            mSameUUID = false;
            if (!sameUUID && mCard.Type != CardType.CURSE && mCard.Type != CardType.STATUS && DungeonManager.Inst.Player.GetPower("MasterRealityPower") != null)
            {
                mCard.Upgrade();
            }
        }

        public void Init()
        {
            
        }

        public  UniTask<bool> Update()
        {
            if (this.Duration >= 0.5f)
            {
                if (mNumCards < 6)
                {
                    for (int i = 0; i < this.mNumCards; i++)
                    {
                        DungeonManager.EffectManager.Play(new ShowCardAndAddToDiscardEffect(mCard.Clone(mSameUUID)));
                    }
                }

                return UniTask.FromResult(true);
            }

            return UniTask.FromResult(false);
        }
    }
}